src/ops/attention.cc

```diff
@@ -762,9 +762,16 @@ bool MultiHeadAttention::measure_operator_cost(
       (float const *)sim->allocate(sub_key.get_volume(), DT_FLOAT);
   float const *value_ptr =
       (float const *)sim->allocate(sub_value.get_volume(), DT_FLOAT);
-  float const *weight_ptr = (float const *)sim->allocate(num_weights, DT_FLOAT);
+  cost_metrics.inputs_memory = static_cast<size_t>(sim->offset);
```

Comment:
Would it be good to abstract this part out? Seems like this needs to be added/copy-and-pasted to a bunch of operators. Not saying it should be, or even that it needs to be done in this PR, just something to think about in the long term

Reply:
Yes, I'm going to factor it out.

Reply:
The new factoring out seems like an improvement but there's still a lot of repetition 

For example, `BatchNorm` contains
```cpp
  sim->free_all();
  float *input_ptr = (float *)sim->allocate(sub_input.get_volume(), DT_FLOAT);
  assert(input_ptr != NULL);
  cost_metrics.inputs_memory += cost_metrics.total_mem_diff_from(sim->offset);

  float *output_ptr = (float *)sim->allocate(sub_output.get_volume(), DT_FLOAT);
  assert(output_ptr != NULL);
  cost_metrics.outputs_memory += cost_metrics.total_mem_diff_from(sim->offset);

  float *bias_ptr = (float *)sim->allocate(output_c, DT_FLOAT);
  assert(bias_ptr != NULL);
  float *scale_ptr = (float *)sim->allocate(output_c, DT_FLOAT);
  assert(scale_ptr != NULL);
  cost_metrics.weights_memory += cost_metrics.total_mem_diff_from(sim->offset);
  ```
  while `Conv2D` contains
  ```cpp
  sim->free_all();
  float *input_ptr = (float *)sim->allocate(sub_input.get_volume(), DT_FLOAT);
  assert(input_ptr != NULL);
  cost_metrics.inputs_memory += cost_metrics.total_mem_diff_from(sim->offset);

  float *output_ptr = (float *)sim->allocate(sub_output.get_volume(), DT_FLOAT);
  assert(output_ptr != NULL);
  cost_metrics.outputs_memory += cost_metrics.total_mem_diff_from(sim->offset);

  float *weight_ptr = (float *)sim->allocate(
      (size_t)output_c * input_c * kernel_h * kernel_w / groups, DT_FLOAT);
  assert(weight_ptr != NULL);
  float *bias_ptr = (float *)sim->allocate(output_c, DT_FLOAT);
  assert(bias_ptr != NULL);
  cost_metrics.weights_memory += cost_metrics.total_mem_diff_from(sim->offset);
  ```
They're almost identical, and this functionality is duplicated across all of the different operators. I don't have a clear vision of how I'd do it and it doesn't necessarily have to be done in this PR, but it would be nice to find some way to factor this out to make it easier to maintain

Reply:
Yes, I agree - we need to have a better way to redesign the op implementation and the new implementation should handle this more elegantly.